### DISCUSSION TITLE
Correct way to abort waiting in `waitFor`?

### DISCUSSION THREAD

**Discussion by crishoj** [statelyai/xstate#4457] [Q&A] [answered]
I'm signaling `otherMachine` from a service in machine A, by sending a `MOVE` signal and then using `waitFor` to wait for `otherMachine` to reach the `Ready` state.

If `otherMachine` in the meantime reaches a failure state, I'd like to abort `waitFor` immediately, instead of waiting until the `timeout` passed to `waitFor` has passed.

The following is what I came up with, but I am unsure if this is the best way:

```ts
services: {
    moveItem: (ctx) => new Promise<void>((resolve, reject) => {
        ctx.otherMachine.send({type: "MOVE", data: {item: ctx.item}})
        waitFor(ctx.otherMachine, (state) => {
            if (state.matches('Moving.Failure')) reject(new Error('Move failed unrecoverably'))
            return state.matches('Ready')
        }, {timeout: 60000}).then(() => resolve())
    })
}
```

I also tried throwing in `waitFor`, but this led to an error along the lines of "Missing onError handler and/or unhandled exception/promise rejection for invocation".

**Comment by Andarist** [comment, marked as answer]
Waiting for both states that you are interested in (`'Ready'` and `'Moving.Failure'`) sounds like a very good solution if you want to stick with `waitFor`. Alternatively, in v5 you will be able to send `self` together with your `'MOVE'` event and the recipient should be able to "respond" later directly to your "requester" - so you wouldn't even need to use `waitFor`. YMMV though.

**Reply by crishoj** [reply]
Thank you, @Andarist.

In case of waiting for both states, would I then be able to conditionally transition to a new state in `onDone` using a guard, or is only a single target allowed for `onDone`?

This is what I have in mind:

```ts
onDone: [
    {target: "StateA", cond: "someGuard"},
    {target: "StateB"}
]
```

The V5 solution sounds promising as well!

**Reply by crishoj** [reply]
@Andarist we have migrated our actor logic to v5. Can you refer to an example of passing `self` to a child actor that "responds" done/failed events back to the requester?

**Reply by davidkpiano** [reply]
@crishoj You can pass `self` into input:

```ts
// in setup(…)
actors: {
  someActor: fromPromise(async ({ input }) => {
    input.parent; // actor ref
  })
}

// in machine
invoke: {
  src: 'someActor',
  input: ({ self }) => ({ parent: self })
  // ...
}
```

**Reply by ichunge** [reply]
学习了 get!
